### ISSUE TITLE
PowerShell 7.2 Runbook does not format webhook data correctly

### ISSUE THREAD

**Issue by CCHV-GitHub** [PowerShell/PowerShell#19781]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

When a powershell runbook 7.2 receives WebhookData from an alert it does not store it as JSON, it looks like it stores it as a string and thus cannot convert it from JSON into something usable in powershell.

This works fine with a 5.1 runbook the webhook data is received correctly as JSON.

### Expected behavior

```console
Runbook Webhook data is received and formatted as JSON
```

### Actual behavior

```console
Runbook Webhook data is received and but formatted as String
```

### Error details

```console
No error, just a variable with an unusable string.
```

### Environment data

```powershell
This uses Azure Automation Runbook 7.2

Name                           Value

----                           -----

PSVersion                      7.2.0

PSEdition                      Core

GitCommitId                    7.2.0

OS                             Microsoft Windows 10.0.17763

Platform                       Win32NT

PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}

PSRemotingProtocolVersion      2.3

SerializationVersion           1.1.0.1

WSManStackVersion              3.0
```

**Comment by Seidlm** [comment]
is there any solution for this, this is a showstopper right now.

**Comment by Jan1503** [comment]
Same problem here, major showstopper! Need to be fixed asap!

**Comment by Seidlm** [comment]
i got Feedback from Product Team, this should be fixed with the November sprint,

**Comment by fred31330** [comment]
To the Product Team: I wasted several hours trying to figure out what was going wrong with my webhook before I finally stumbled upon this issue. You need to communicate whenever there is such major issue that is breaking our scripts in production, and not letting your users down like that! This is very frustrating, above all when this is not the first time it happens.
Sorry for the rant, but I had to let it out.

**Comment by Seidlm** [comment]
@fred31330 took me also some hours to figure out. It is frustrating, but soon it will be fixed.

**Comment by fred31330** [comment]
When is the new sprint released with this fix?

**Comment by PeterJamesLawson** [comment]
@Seidlm do you know if the product team will release this fix in the November release?

**Comment by Seidlm** [comment]
They told me so... :/

**Comment by MichaelSeidl1982** [comment]
Hi, haven't received any feedback from Microsoft, so seems still unsolved

**Comment by Seidlm** [comment]
Hi, still no feedback from MS, sor seems not solved

**Comment by fred31330** [comment]
What a pity!
@Seidlm Would you mind asking the product team again?
Thank you!

**Comment by Seidlm** [comment]
hi, finally received an update

"This bug is triaged to be picked up in March sprint with ETA of 3/30 to be available."

**Comment by ITJoeSchmo** [comment]
We have been using this work-around for the issue for a few days, it may do the trick for you: 

`# 1) Quote unquoted top-level keys
#    Looks for { or , followed by a wordlike sequence up to the colon,
#    ensures that it's not already quoted (negative lookahead (?!"))
$WebhookData = $WebhookData -replace '(?<=[\{,])\s*([A-Za-z0-9_-]+)\s*(?=:)(?!")', '"$1"'

# 2) Quote unquoted top-level values
#    Looks for the colon *after* a newly quoted key ("Key":),
#    ensures next char isn’t a brace or a quote (so we skip nested objects or already-quoted strings).
#    Captures up to the next comma or closing brace, then inserts quotes around it.
$WebhookData = $WebhookData -replace '(?<="):\s*(?!")([^,"{}]+)(?=[,}])', ':"$1"'`

**Comment by jan-swinkels** [comment]
> hi, finally received an update
> 
> "This bug is triaged to be picked up in March sprint with ETA of 3/30 to be available."

Today it is passed 3/30/2025 .. Still not working ...
Is it for Microsoft that difficult to fix this ??

Any other people who tested the " solution" from Microsoft and found it was solved by Microsoft??

**Comment by ibitforcaa** [comment]
Any movement on this issue? 
I took a working runbook using PS 7.2 and moved it to a new Automation account attempted to use PS 7.2 in the new automation account. None of my webhook data was usable, down graded the runbook to PS 5 and the runbook works well, all webhook data is formatted correctly. 
I am unsure why older PS 7.2 webhook data is formatted correctly but when a new PS 7.2 runbook is created the webhook data is unusable.

**Comment by Seidlm** [comment]
I received an update from the product team, there is no plan right now when and if this will be fixed.

**Comment by jan-swinkels** [comment]
> [Seidlm](https://github.com/Seidlm) received an update from the product team, there is no plan right now when and if this will be fixed.

OK .. Bad News ...
Still no answer from Microsoft why Basic functionality like WebHooks are not working as expected in PS 7 

Good News -> Webhook on PS5 is working.
My Solution -> Webhook to PS5 script    ... only one thing this PS 5 script will do is LAUNCH my PS 7 script !!!!
( PS5 script will launch PS 7 script as the PS7 WebHook bug ... )

**Comment by fred31330** [comment]
> I received an update from the product team, there is no plan right now when and if this will be fixed.

Well first of all @Seidlm, thanks for acting as our relay to the product team. 

But honestly, I’m really frustrated at this point. The bug we’ve all been waiting on for so long was supposed to be fixed in the November sprint. Then it got pushed to March. Now we're being told there’s _no plan at all_ to fix it — not even a vague idea of when or if it’ll ever happen?

This isn’t just disappointing — it’s disruptive. As a PowerShell developer, I’ve built work around the expectation that this would be resolved based on the timelines the product team has given. Shifting priorities are one thing, but moving the goalposts repeatedly and then removing them entirely without warning makes it incredibly hard to plan or deliver anything reliably.

To the product team:
If something’s not going to be fixed, say so upfront! Don’t string people along for months and then shrug it off. That’s not how you build trust or credibility with those depending on your work.

**Comment by Seidlm** [comment]
i know it is frustrating, i am used of it and still using ps5 for azure automation.

**Comment by jformacek** [comment]
I'm frustrated the same way - PS7.2 now being recommended runtime and webhook data unusable - no improvement on the horizon.
It would be fair from MS that product group at least published guidance here how to work with WebhookData input formatted as it is now?

**Comment by BenjaminZ1** [comment]
It's really frustrating and holds us back. This is the main reason why all our runbooks are still in the Powershell 5.1 runtime environment.

**Comment by JaibeHyne** [comment]
Damn, this is still an issue in Nov 2025. Hybrid Workers using PowerShell 7.2-7.5 have this issue. The WebhookData is received as a String. Very frustrating wasting all this time.

**Comment by JaibeHyne** [comment]
> We have been using this work-around for the issue for a few days. The problem is that the strings in the JSON are not being wrapped in quotes. It may end up translating values that would be int to string, but you can recast it in the script if needed. It may do the trick for you:
> 
> ```
> # 1) Quote unquoted top-level keys
> #    Looks for { or , followed by a wordlike sequence up to the colon,
> #    ensures that it's not already quoted (negative lookahead (?!"))
> $WebhookData = $WebhookData -replace '(?<=[\{,])\s*([A-Za-z0-9_-]+)\s*(?=:)(?!")', '"$1"'
> 
> # 2) Quote unquoted top-level values
> #    Looks for the colon *after* a newly quoted key ("Key":),
> #    ensures next char isn’t a brace or a quote (so we skip nested objects or already-quoted strings).
> #    Captures up to the next comma or closing brace, then inserts quotes around it.
> $WebhookData = $WebhookData -replace '(?<="):\s*(?!")([^,"{}]+)(?=[,}])', ':"$1"'
> ```

Thanks, works a treat. I just had to add some \\t tab removals and we're good to go.

**Comment by benjaminxie** [comment]
I'm in a similar situation. I hit an error before a single line of my runbook runs. If you're not sure what JaibeHyne means by "the WebhookData is received as a String," the following is an explanation. 

My runbook job fails with an error that is actually split among five entries in the job's error tab and they all have color escape code related gibberish. I've copied them below on their own lines:

`[31;1mParserError: [0m

[31;1m[36;1mLine |[0m

[31;1m[36;1m[36;1m 1 | [0m … agent019c7faa-3d6f-446e-bbe5-11a0a74d164b.ps1 {'WebhookName'[36;1m:[0m 'Alert1 …[0m

[31;1m[36;1m[36;1m[0m[36;1m[0m[36;1m | [31;1m ~[0m

[31;1m[36;1m[36;1m[0m[36;1m[0m[36;1m[31;1m[31;1m[36;1m | [31;1mUnexpected token ':' in expression or statement.[0m`

Based these errors, it looks like my code is being run similar to this

```
> ./helloWorld.ps1 {"WebhookName": "Alert1764881567264"}                                                                                                     ParserError: 
Line |
   1 |  ./helloWorld.ps1 {"WebhookName": "Alert1764881567264"}
     |                                 ~
     | Unexpected token ':' in expression or statement.
```

So I'm basically getting a syntax error from the powershell terminal because the webhook data isn't being passed properly.

What's worse is that it looks like I can't even use Powershell 5.1 like the people above. I created a runbook that uses 5.1 and I get the error `Failed
This runbook type isn't supported by Linux hybrid workers.`.

Edit: Nevermind, using Powershell 5.1 is possible. I just have to remember to add an environment variable called powershell_5_1_path to /etc/environment. And it can even just point right back to whatever powershell_7_2_path pointed to. (Source: https://learn.microsoft.com/en-us/azure/automation/automation-hrw-run-runbooks?tabs=sa-mi#extension-based-hybrid-workers-1)

**Comment by bmackalicious** [comment]
Oof, been spending all afternoon trying to troubleshoot this only to find this open issue... :(

**Comment by benjaminxie** [comment]
@bmackalicious The solution I ended up using was to use a sort of parent-child pair of runbooks. 

Just to reiterate, the problem I was having was that my runbook running on a hybrid worker wouldn't be passed the webhook data properly. On the other hand, my runbook running in the Azure sandbox worked fine (but I needed the hybrid worker due to networking reasons).

So back to the parent-child runbooks. The parent runbook runs in the Azure sandbox and takes only the webhook as an argument. All it does is parse the webhook and call the child runbook, passing the necessary data from the webhook (my child runbook only took string arguments). The child runbook, no longer needing to be passed the webhook, can now run properly on a hybrid worker. 

The cmdlet I used was Start-AzAutomationRunbook. You'll also need to give your managed identity the proper role. Automation Operator worked for me.

To recap, if you have a runbook that takes a webhook and runs on a hybrid worker, split it up into two runbooks. One that runs in the Azure sandbox and parses the webhook, and another that runs on a hybrid worker and does the real work. Hope this helps!

**Comment by mpolselli** [comment]
Pinging the recent participants as I am hoping this is helpful.
@Seidlm @benjaminxie @rakheshster @Jan1503 @jformacek @fred31330 @JaibeHyne @BenjaminZ1 @jan-swinkels @ibitforcaa @ITJoeSchmo @bmackalicious 

I ended up with a similar workaround to @benjaminxie, the difference being that I have one Azure runbook that has a webhook that acts as the orchestrator and can agnostically call any of my other runbooks and provide parameters to them. The JSON payload needs to follow a specific format, but I find this makes things easier to manage and organize. Hopefully this is a cleaner workaround until the product group provides a proper fix.

Azure Runbook
```powershell
param(
    $WebhookData
)

$connectAzAccount = @{
    Identity     = $true
    Environment  = 'environment'
    Tenant       = $tenantId
    Subscription = $subscriptionId
}
Connect-AzAccount @connectAzAccount | Out-Null

$body = $WebhookData.RequestBody | ConvertFrom-Json -AsHashtable

$hybridRunbookParams = @{}

$body.Parameters.GetEnumerator() | ForEach-Object {
    $hybridRunbookParams[$PSItem.Name] = $PSItem.Value
}

$params = [PSCustomObject]@{
    RunbookName = $body.RunbookName
    Wait        = $body.Wait
    Parameters  = $hybridRunbookParams
}

$startAzAutomationRunbook = @{
    AutomationAccountName = 'AccountName'
    ResourceGroupName     = 'RGName'
    Name                  = $params.RunbookName
    Parameters            = $params.Parameters
    Wait                  = $params.Wait
    RunOn                 = 'HybridWorkerGroup'
}
Start-AzAutomationRunbook @startAzAutomationRunbook
```

JSON payload 1
```json
{
    "RunbookName": "test123",
    "Wait": true,
    "Parameters": {
        "Param1": "value1",
        "Param2": "value2",
    }
}
```

JSON payload 2
```json
{
    "RunbookName": "OnboardUser",
    "Wait": false,
    "Parameters": {
        "Username": "j_doe",
        "Mailbox": true,
        "DeskNumber": 42
    }
}
```

JSON payload 3
```json
{
    "RunbookName": "TestBook",
    "Wait": false,
}
```

JSON schema
```json
{
    "$schema": "https://json-schema.org",
    "type": "object",
    "properties": {
        "RunbookName": {
            "type": "string"
        },
        "Wait": {
            "type": "boolean"
        },
        "Parameters": {
            "type": [
                "object",
                "null"
            ],
            "properties": {
                "Param1": {
                    "type": [
                        "integer",
                        "string",
                        "boolean",
                        "object"
                    ]
                },
                "Param2": {
                    "type": [
                        "integer",
                        "string",
                        "boolean",
                        "object"
                    ]
                },
                "ParamEtc": {
                    "type": [
                        "integer",
                        "string",
                        "boolean",
                        "object"
                    ]
                }
            }
        }
    },
    "required": [
        "RunbookName",
        "Wait"
    ]
}
```
